packages/esm-service-queues-app/src/queue-table/queue-table.component.tsx

```diff
@@ -0,0 +1,216 @@
+
+  const visitedQueueEntriesByLocationAndQueueAndStatus = visitQueueEntriesByLocation
+    ?.filter((entry) => !currentQueueUuid || entry.queueEntry.queue.uuid == currentQueueUuid)
+    ?.filter((entry) => !currentStatusUuid || entry.queueEntry.status.uuid == currentStatusUuid);
+
+  if (isLoading || !allowedStatuses?.length) {
+    return <>Loading....</>;
+  }
+
+  const handleServiceChange = ({ selectedItem }: { selectedItem: QueueServiceInfo }) => {
+    updateSelectedServiceUuid(selectedItem.uuid);
+  };
+
+  return (
+    <div className={styles.container}>
+      <div className={styles.headerContainer}>
+        <div className={!isDesktop(layout) ? styles.tabletHeading : styles.desktopHeading}>
+          <h4>{t('patientsCurrentlyInQueue', 'Patients currently in queue')}</h4>
+        </div>
+      </div>
+      <div className={styles.filterContainer}>
+        <Dropdown
+          id="serviceFilter"
+          titleText={t('showPatientsWaitingFor', 'Show patients waiting for') + ':'}
+          label={currentQueue?.name ?? t('all', 'All')}
+          type="inline"
+          items={[{ display: `${t('all', 'All')}` }, ...queues]}
+          itemToString={(item) => (item ? item.display : '')}
+          onChange={handleServiceChange}
+          size="sm"
+        />
+      </div>
+      <Tabs
+        selectedIndex={currentStatusIndex}
+        onChange={({ selectedIndex }) => setCurrentStatusUuid(allowedStatuses[selectedIndex].uuid)}
+        className={styles.tabs}>
+        <TabList style={{ paddingLeft: '1rem' }} aria-label="Outpatient tabs" contained>
+          {allowedStatuses?.map((s) => <Tab key={s?.uuid}>{s?.display}</Tab>)}
+        </TabList>
+        <TabPanels>
+          {allowedStatuses?.map((s) => (
+            <TabPanel key={s?.uuid} style={{ padding: 0 }}>
+              <QueueTable
+                visitedQueueEntriesByLocationAndQueueAndStatus={visitedQueueEntriesByLocationAndQueueAndStatus}
+              />
+            </TabPanel>
+          ))}
+        </TabPanels>
+      </Tabs>
+    </div>
+  );
+};
+
+interface QueueTableProps {
+  visitedQueueEntriesByLocationAndQueueAndStatus: VisitQueueEntry[];
+}
+
+function QueueTable({ visitedQueueEntriesByLocationAndQueueAndStatus }: QueueTableProps) {
+  const { t } = useTranslation();
+  const { queueTableColumns } = useConfig<ConfigObject>();
+  const [currentPageSize, setPageSize] = useState(10);
+  const pageSizes = [10, 20, 30, 40, 50];
+  const {
+    goTo,
+    results: paginatedQueueEntries,
+    currentPage,
+  } = usePagination(visitedQueueEntriesByLocationAndQueueAndStatus, currentPageSize);
+
+  const headers = queueTableColumns.map((column) => ({ header: t(column.headerI18nKey), key: column.id }));
+  const rows =
+    paginatedQueueEntries?.map((queueEntry) => {
+      const row = {};
+      queueTableColumns.forEach((conf) => {
+        row[conf.id] = <ExtensionSlot name={conf.extensionSlotName} state={{ queueEntry }} />;
```

Comment:
Yikes! This is stress-testing the extension system! I was imagining this working as something like a slot-per-column, or, better yet, a simple mapping function rather than a slot-per-cell. I guess we'll see how this works...

Reply:
For reference, a 20-person queue with just 3 columns is now 60 separate slots with 60 separate extensions.

Reply:
Is that true?  If so, I agree we may need an alternative.  Ultimately, we just want to be able to have a list of columnDefinitions that can be configured with a value function or component that takes in a queueEntry as an argument/prop.  If there was a straightforward way to register functions or components with a lookup key, and then to refer to these with this key to use when rendering, that's what we need.

It isn't entirely similar, but I'd think this kind of thing could also eliminate the kind of hard-coding that one finds in places [like this](https://github.com/openmrs/openmrs-esm-patient-management/blob/main/packages/esm-patient-registration-app/src/patient-registration/field/custom-field.component.tsx#L16) and [this](https://github.com/openmrs/openmrs-esm-patient-management/blob/main/packages/esm-patient-registration-app/src/patient-registration/field/field.component.tsx#L31).  The similarity is that there is a need to support end-user configuration via a string value, and then have this string value determine what actual component is rendered.

Reply:
@ibacher I did some rough [profiling ](https://github.com/openmrs/openmrs-esm-patient-management/commit/069738711151750b9edcdddea377b8dc7bb4a6e3) work. Using a table with 10 rows and 3 columns on my computer, I measure the time to render the cells using the components directly vs using `<ExtensionSlot>` . The direct way has 1ms of time lapse between the `useEffect(..., [])` calls of the first and the last rendered cells, while via `<ExtensionSlot>` the time lapse is 20ms. So... technically 20x worse, but still small enough to be acceptable?

Reply:
Just seeing this thread now, it wasn't visible when I was looking at the PR itself... did we move away from this approach?  I agree that having a slot-per-cell seems like a lot, but don't necessarily want to push for a full refactor now if we want to just see how it works in this case.  

One goal on my end would be keep the code a simple as possible (the key being "as possible", I know there will be some complexity) from a extension and readability perspective for what I suspect will be the 80% use case for most implementions, just to use the default columns.

Reply:
I don't love this approach., but I'm keeping it as is for now (unless others have objection). It's ok performance-wise, but developer experience isn't great as we need somewhat bloated JSON to define the columns. This approach also doesn't address how to go about having the table show different columns at different places. (For example, we might want a "current queue" column when we are looking at queue entries from all queues, but can hide the column if we're only looking at a specific queue.)